testsuite/devserver.go

```diff
@@ -0,0 +1,356 @@
+	// Which version to download, by default the latest version compatible with the SDK will be downloaded.
+	// Acceptable values are specific release versions (e.g v0.3.0), "default", and "latest".
+	Version string
+	// Destination directory or the user temp directory if unset.
+	DestDir string
+}
+
+// Where to find an executable. Can be a path or cached download.
+type EphemeralExe struct {
+	// Existing path on the filesystem for the executable.
+	ExistingPath string
+	// Download the executable if not already there.
+	CachedDownload CachedDownload
+}
+
+// Configuration for the dev server.
+type DevServerOptions struct {
+	// Path to executable or download info - defaults to cached download.
+	Exe EphemeralExe
+	// Client options used to create a client for the dev server.
+	// The provided Namespace or the "default" namespace is automatically registered on startup.
+	// If HostPort is provided, the host and port will be used to bind the server, otherwise the server will bind to
+	// localhost and obtain a free port.
+	ClientOptions *client.Options
+	// SQLite DB filename if persisting or non-persistent if none.
+	DBFilename string
+	// Whether to enable the UI.
+	EnableUI bool
+	// Log format - defaults to "pretty"
+	LogFormat string
+	// Log level - defaults to "warn"
+	LogLevel string
+	// Additional arguments to the dev server.
+	ExtraArgs []string
+}
+
+type DevServer interface {
+	// Stop the running server and wait for shutdown to complete. Error is propagated from server shutdown.
+	Stop() error
+	// Get a connected client, configured to work with the dev server.
+	Client() client.Client
+}
+
+// Temporal CLI based DevServer
+type devServer struct {
+	cmd    *exec.Cmd
+	client client.Client
+}
+
+// StartDevServer starts a Temporal CLI dev server process. This may download the server if not already downloaded.
+func StartDevServer(ctx context.Context, options *DevServerOptions) (DevServer, error) {
+	// Accept nil because all options are "optional".
+	if options == nil {
+		options = &DevServerOptions{}
+	}
+	clientOptions := options.clientOptionsOrDefault()
+
+	exePath, err := downloadIfNeeded(options, clientOptions.Logger)
+	if err != nil {
+		return nil, err
+	}
+
+	if clientOptions.HostPort == "" {
+		// Make sure this is done after downloading to reduce the chance (however slim) that the free port would be used
+		// up by the time the download completes.
+		clientOptions.HostPort, err = getFreeHostPort()
+		if err != nil {
+			return nil, err
+		}
+	}
+	host, port, err := net.SplitHostPort(clientOptions.HostPort)
+	if err != nil {
+		return nil, fmt.Errorf("invalid HostPort: %w", err)
+	}
+
+	args := []string{
+		"server",
+		"start-dev",
+		"--ip", host, "--port", port,
+		"--namespace", clientOptions.Namespace,
+		"--dynamic-config-value", "frontend.enableServerVersionCheck=false",
+		"--dynamic-config-value", "system.enableEagerWorkflowStart=true",
+		"--dynamic-config-value", "system.enableEagerWorkflowStart=true",
```

Comment:
Why are we enabling eager workflow start on the dev server here? The dev server should match the non-dev server IMO. When we enable the non-dev server form of this we can enable the dev server form. They can opt in to this just like they do on the real server.